llvm/lib/Target/X86/X86RegisterInfo.cpp

```diff
@@ -158,6 +158,10 @@ X86RegisterInfo::getLargestLegalSuperClass(const TargetRegisterClass *RC,
     case X86::GR16RegClassID:
     case X86::GR32RegClassID:
     case X86::GR64RegClassID:
+    case X86::GR8_NOREX2RegClassID:
+    case X86::GR16_NOREX2RegClassID:
+    case X86::GR32_NOREX2RegClassID:
+    case X86::GR64_NOREX2RegClassID:
```

Comment:
It's not clear to me when we need to distinguish `X86::GR8_NOREX2RegClassID` from `X86::GR8RegClassID` and when not. We have some other places, e.g., [here ](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86InsertPrefetch.cpp#L87-L91)that using `X86::GRxxRegClassID`, shouldn't need to update with `X86::GRxx_NOREX2RegClassID`
Besides, we also have some place to `createVirtualRegister(&X86::GRxxRegClass)`, should they need to update too?

Reply:
We have comments in X86BaseInfo.h about when we need to distinguish them. `PREFETCH` instructions are in map TB and they can use r16-r31, so `X86::GRxxRegClassID` does not need to be updated.

Yes, some of `createVirtualRegister(&X86::GRxxRegClass)` needs to be updated if the instruction can not encode r16-r31.

Reply:
Will you update in this patch or a follow up?

Reply:
Didn't look through one by one, but [here](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86DomainReassignment.cpp#L186-L189) seems have risk since it itrates different instructions.

Reply:
There is no `createVirtualRegister(&X86::GRxxRegClass)` for instructions that can not encode EGPR so far. If there is any in the future, we will directly use `createVirtualRegister(&X86::GRxx_NOREX2RegClass)` at those places.

Reply:
What's the risk?
```
    Register Reg = MRI->createVirtualRegister(
        TII->getRegClass(TII->get(DstOpcode), 0, MRI->getTargetRegisterInfo(),
                         *MBB->getParent()));
    MachineInstrBuilder Bld = BuildMI(*MBB, MI, DL, TII->get(DstOpcode), Reg);
```
The code calls `getRegClass` to get the register class and then build the machine instruction with the same opcode. It looks safe to me.

In X86,   (If I remember correctly) pseudo instruction `COPY` is either a `MOV` or `KMOV`. Both of them can encode r16-31.